common/client/multi_node.go

```diff
@@ -545,45 +556,187 @@ func (c *multiNode[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OP
 	return n.RPC().SendEmptyTransaction(ctx, newTxAttempt, seq, gasLimit, fee, fromAddress)
 }
 
+type sendTxResult struct {
+	Err        error
+	ResultCode SendTxReturnCode
+}
+
+// broadcastTxAsync - creates a goroutine that sends transaction to the node. Returns false, if MultiNode is Stopped
+func (c *multiNode[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, RPC_CLIENT]) broadcastTxAsync(ctx context.Context,
+	n SendOnlyNode[CHAIN_ID, RPC_CLIENT], tx TX, txResults chan sendTxResult, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	txErr := n.RPC().SendTransaction(ctx, tx)
+	c.lggr.Debugw("Node sent transaction", "name", n.String(), "tx", tx, "err", txErr)
+	resultCode := c.classifySendTxError(tx, txErr)
+	if resultCode != Successful && resultCode != TransactionAlreadyKnown {
+		c.lggr.Warnw("RPC returned error", "name", n.String(), "tx", tx, "err", txErr)
+	}
+
+	// we expected txResults to have sufficient buffer, otherwise we are not interested in the response
+	// and can drop it
+	select {
+	case txResults <- sendTxResult{Err: txErr, ResultCode: resultCode}:
+	default:
+	}
+}
+
+func fanOut[T any](source chan T, destinations ...chan T) {
+	for t := range source {
+		for _, dest := range destinations {
+			dest <- t
+		}
+	}
+
+	for _, dest := range destinations {
+		close(dest)
+	}
+}
+
+// collectTxResults - waits for sufficient number of sendTxResult to determine result of transaction submission.
+// We balance the waiting time and the number of collected results. Our target is replies from 70% of nodes,
+// but we won't wait longer than sendTxSoftTimeout since the first reply to avoid waiting
+// for a timeout from slow/unhealthy nodes.
+func (c *multiNode[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, RPC_CLIENT]) collectTxResults(ctx context.Context, tx TX, txResults <-chan sendTxResult) (map[SendTxReturnCode][]error, error) {
+	// combine context and stop channel to ensure we stop, when signal received
+	ctx, cancel := c.chStop.Ctx(ctx)
+	defer cancel()
+	const quorum = 0.7
+	requiredResults := int(math.Ceil(float64(len(c.nodes)) * quorum))
+	errorsByCode := map[SendTxReturnCode][]error{}
+	var softTimeoutChan <-chan time.Time
+	var resultsCount int
+	for {
+		select {
+		case <-ctx.Done():
+			c.lggr.Debugw("Failed to collect of the results before context was done", "tx", tx, "errorsByCode", errorsByCode)
+			return nil, ctx.Err()
+		case result := <-txResults:
+			errorsByCode[result.ResultCode] = append(errorsByCode[result.ResultCode], result.Err)
+			resultsCount++
+			if result.ResultCode == Successful || resultsCount >= requiredResults {
+				return errorsByCode, nil
+			}
+		case <-softTimeoutChan:
```

Comment:
What if we have 0 responses so far? This seems to be half of the value for a standard `QueryTimeout`

Reply:
If there is 0 responses, softTimeoutChan is nil and won't fire. We initialize it when we receive first reply